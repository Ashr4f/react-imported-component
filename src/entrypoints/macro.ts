// @ts-ignore
import { createMacro } from 'babel-plugin-macros';
import { createTransformer } from '../babel/babel';
import { assignImportedComponents } from '../loadable/loadable';

function getMacroType(tagName: string) {
  switch (tagName) {
    case 'importedModule':
    case 'imported':
    case 'lazy':
    case 'useImported':

    case 'ImportedModule':
    case 'ImportedComponent':
      return 'react-imported-component';

    case 'assignImportedComponents':
      return 'react-imported-component/boot';

    default:
      return false;
  }
}

function macro({ references, state, babel }: any) {
  const { types: t } = babel;
  const fileName = state.file.opts.filename;

  const imports: Record<string, string[]> = {};
  const transformer = createTransformer(babel, true);

  Object.keys(references).forEach((tagName: string) => {
    const origin = getMacroType(tagName);
    if (origin) {
      imports[origin] = imports[origin] || [];
      imports[origin].push(tagName);
      const tags = references[tagName];
      tags.forEach((tag: any) => {
        const expression = tag.parentPath;

        if (t.isCallExpression(expression)) {
          transformer.traverse(expression, fileName);
        }
      });
    }
  });

  addReactImports(babel, state, imports);
  transformer.traverse(state.file.path, fileName);
  transformer.finish(state.file.path.node, fileName);
}

function addReactImports(babel: any, state: any, importsGroups: Record<string, string[]>) {
  const { types: t } = babel;

  return Object.keys(importsGroups)
    .map(origin => {
      const imports = importsGroups[origin];

      if (!imports.length) {
        return false;
      }

      const importedImport = state.file.path.node.body.find(
        (importNode: any) => t.isImportDeclaration(importNode) && importNode.source.value === origin
      );

      // Handle adding the import or altering the existing import
      if (importedImport) {
        imports.forEach(name => {
          if (
            !importedImport.specifiers.find((specifier: any) => specifier.imported && specifier.imported.name === name)
          ) {
            importedImport.specifiers.push(t.importSpecifier(t.identifier(name), t.identifier(name)));
          }
        });
      } else {
        state.file.path.node.body.unshift(
          t.importDeclaration(
            imports.map(name => t.importSpecifier(t.identifier(name), t.identifier(name))),
            t.stringLiteral(origin)
          )
        );
      }

      return true;
    })
    .some(Boolean);
}

const neverCallMe: any = () => {
  throw new Error(
    'you have used `react-imported-component/macro` without `babel-plugin-macro` or `react-hot-loader/babel` installed'
  );
};

const lazy: typeof import('../ui/HOC').lazy = neverCallMe;
const imported: typeof import('../ui/HOC').default = neverCallMe;
const importedModule: typeof import('../ui/Module').importedModule = neverCallMe;
const useImported: typeof import('../ui/useImported').useImported = neverCallMe;

const ImportedModule: typeof import('../ui/Module').ImportedModule = neverCallMe;
const ImportedComponent: typeof import('../ui/Component').ImportedComponent = neverCallMe;

export { lazy, imported, importedModule, ImportedModule, ImportedComponent, useImported, assignImportedComponents };

export default createMacro(macro);
