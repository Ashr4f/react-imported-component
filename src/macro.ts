// @ts-ignore
import {createMacro} from "babel-plugin-macros";
import {createTransformer} from "./babel"
import {assignImportedComponents} from './loadable'

function getMacroType(tagName: string) {
  switch (tagName) {
    case "imported":
    case "lazy":
    case "useImported":
      return "react-imported-component";
    case "assignImportedComponents":
      return "react-imported-component/boot";

    default:
      return false;
  }
}

function macro({references, state, babel}: any) {
  const {types: t} = babel;

  const imports: Record<string, string[]> = {};
  const transformer = createTransformer(babel);

  Object
    .keys(references)
    .forEach((tagName: string) => {
      const origin = getMacroType(tagName);
      if (origin) {
        imports[origin] = imports[origin] || [];
        imports[origin].push(tagName);
        const tags = references[tagName];
        tags.forEach((tag: any) => {
          let expression = tag.parentPath;

          if (t.isCallExpression(expression)) {
            transformer.traverse(expression, state.file)
          }
        });
      }
    });

  if (addReactImports(babel, state, imports)) {
    transformer.finish(state.file.path.node);
  }
}

function addReactImports(babel: any, state: any, importsGroups: Record<string, string[]>) {

  const {types: t} = babel;

  return (
    Object
      .keys(importsGroups)
      .map(origin => {
        const imports = importsGroups[origin];

        if (!imports.length) return false;

        const importedImport = state.file.path.node.body.find(
          (importNode: any) =>
            t.isImportDeclaration(importNode) &&
            importNode.source.value === origin
        );

        // Handle adding the import or altering the existing import
        if (importedImport) {
          imports.forEach(name => {
            if (
              !importedImport.specifiers.find(
                (specifier: any) => specifier.imported && specifier.imported.name === name
              )
            ) {
              importedImport.specifiers.push(
                t.importSpecifier(t.identifier(name), t.identifier(name))
              )
            }
          })
        } else {
          state.file.path.node.body.unshift(
            t.importDeclaration(
              imports.map(name =>
                t.importSpecifier(t.identifier(name), t.identifier(name))
              ),
              t.stringLiteral(origin)
            )
          )
        }

        return true;
      })
      .some(Boolean)
  );
}

const neverCallMe: any = () => {
  throw new Error('you have used `react-imported-component/macro` without `babel-plugin-macro` or `react-hot-loader/babel` installed')
};

const lazy: typeof import('./HOC').lazy = neverCallMe;
const imported: typeof import('./HOC').default = neverCallMe;
const useImported: typeof import('./useImported').useImported = neverCallMe;

export {lazy, imported, useImported, assignImportedComponents};

export default createMacro(macro)