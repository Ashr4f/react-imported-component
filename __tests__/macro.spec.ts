const pluginTester = require("babel-plugin-tester");
const plugin = require("babel-plugin-macros");
const prettier = require("prettier");

describe('babel macro', () => {
  pluginTester({
    plugin,
    snapshot: true,
    babelOptions: {
      filename: __filename,
      plugins: ['dynamic-import-node'],
    },
    formatResult(result: string) {
      return prettier.format(result, {trailingComma: "es5"});
    },
    tests: {
      "nothing": "const a = 42;",
      "no usage": `import {lazy} from "../macro";`,
      "lazy": `
        import {lazy} from "../macro";
        const v = lazy(() => import('./a'));
        `,
      "many": `
        import {imported, useImported} from "../macro";        
        const v = imported(() => import('./a'));
        const x = () => useImported(() => import('./b'));
        `,
    },
  });
});