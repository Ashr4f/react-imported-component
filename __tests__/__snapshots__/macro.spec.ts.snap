// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel macro macros many: many 1`] = `
"
import {imported, useImported} from \\"../macro\\";        
const v = imported(() => import('./a'));
const x = () => useImported(() => import('./b'));

      ↓ ↓ ↓ ↓ ↓ ↓

function _interopRequireWildcard(obj) {
  if (obj && obj.__esModule) {
    return obj;
  } else {
    var newObj = {};
    if (obj != null) {
      for (var key in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, key)) {
          var desc =
            Object.defineProperty && Object.getOwnPropertyDescriptor
              ? Object.getOwnPropertyDescriptor(obj, key)
              : {};
          if (desc.get || desc.set) {
            Object.defineProperty(newObj, key, desc);
          } else {
            newObj[key] = obj[key];
          }
        }
      }
    }
    newObj.default = obj;
    return newObj;
  }
}

var importedWrapper = function(marker, realImport) {
  if (typeof __deoptimization_sideEffect__ !== \\"undefined\\") {
    __deoptimization_sideEffect__(marker, realImport);
  }

  return realImport;
};

import { imported, useImported } from \\"react-imported-component\\";
const v = imported(() =>
  importedWrapper(
    \\"imported_-mg71kn_component\\",
    Promise.resolve().then(() => _interopRequireWildcard(require(\\"./a\\")))
  )
);

const x = () =>
  useImported(() =>
    importedWrapper(
      \\"imported_15vaa6j_component\\",
      Promise.resolve().then(() => _interopRequireWildcard(require(\\"./b\\")))
    )
  );

"
`;
